Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean

```diff
@@ -244,6 +253,143 @@ lemma MorphismProperty.topologically_propertyIsLocalAtTarget
+    exact ((RingHom.toMorphismProperty P).arrow_mk_iso_iff <|
+      morphismRestrictStalkMap f (U i) ⟨x, hi⟩).mp <| hf i ⟨x, hi⟩
+
+end Stalkwise
+
+end MorphismProperty
+
+section Restriction
+
+/-- If `P` is a property of scheme morphisms, we may restrict `P` to morphisms with affine target
+to obtain an `AffineTargetMorphismProperty`. -/
+def AffineTargetMorphismProperty.of (P : MorphismProperty Scheme) :
+    AffineTargetMorphismProperty := fun _ _ f ↦ P f
+
+namespace AffineTargetMorphismProperty
+
+/-- Restricting a local at the target morphism property of schemes `P` to morphisms with affine
+target and extending to a global property with `targetAffineLocally` yields `P` again,
+if `P` is local at the target. -/
+lemma targetAffineLocally_of_eq_of_isLocalAtTarget
+    (P : MorphismProperty Scheme) (hP : PropertyIsLocalAtTarget P) :
+    targetAffineLocally (of P) = P := by
+  ext X Y f
+  constructor
+  · intro hf
+    simp only [targetAffineLocally, Subtype.forall] at hf
+    let 𝒰 : Y.OpenCover := Y.affineCover
+    apply hP.of_openCover f 𝒰
+    intro i
+    have hiao : IsAffineOpen (Scheme.Hom.opensRange (𝒰.map i)) :=
+      AlgebraicGeometry.isAffineOpen_opensRange _
+    letI : P.RespectsIso := hP.RespectsIso
+    rw [← P.arrow_mk_iso_iff <| morphismRestrictOpensRange f (𝒰.map i)]
+    exact hf (Scheme.Hom.opensRange (𝒰.map i)) hiao
+  · intro hf ⟨U, hU⟩
+    exact hP.restrict _ _ hf
+
+/-- The restriction of a morphism property of schemes that is local at the target to morphisms
+with affine target, is local. -/
+lemma of_isLocal_of_isLocalAtTarget (P : MorphismProperty Scheme)
+    (hP : PropertyIsLocalAtTarget P) : (of P).IsLocal where
+  RespectsIso := by
+    apply AffineTargetMorphismProperty.respectsIso_mk
+    · intro X Y Z e f _ hf
+      apply hP.RespectsIso.precomp e f hf
+    · intro X Y Z e f _ hf
+      apply hP.RespectsIso.postcomp e f hf
+  toBasicOpen {X Y} _ f hf := hP.restrict f _
+  ofBasicOpenCover {X Y} _ f s hs hf := by
+    apply ((hP.openCover_TFAE f).out 0 5).mpr
+    let U (r : s) : Opens Y.carrier := Y.basicOpen r.val
+    have hiao : IsAffineOpen (⊤ : Opens Y.carrier) := isAffineOpen_top Y
+    have hU : iSup U = ⊤ := by
+      erw [hiao.basicOpen_union_eq_self_iff]
+      exact hs
+    use s, U, hU, hf
+
+lemma stableUnderBaseChange_of_stableUnderBaseChangeOnAffine_of_isLocalAtTarget
+    (P : MorphismProperty Scheme) (hP₁ : PropertyIsLocalAtTarget P)
+    (hP₂ : (of P).StableUnderBaseChange) :
+    P.StableUnderBaseChange := by
+  rw [← targetAffineLocally_of_eq_of_isLocalAtTarget P hP₁]
+  apply (of_isLocal_of_isLocalAtTarget P hP₁).stableUnderBaseChange
+  exact hP₂
+
+end AffineTargetMorphismProperty
+
+open AffineTargetMorphismProperty
+
+/-- A morphism property of schemes is `StableUnderAffineBaseChange` if
+the base change along a morphism of affine schemes of a morphism with affine target is affine. -/
+def StableUnderAffineBaseChange (P : MorphismProperty Scheme) : Prop :=
```

Comment:
Is this just `(of P).StableUnderBaseChange`?

Reply:
Yes, I added this to hide the interaction with `of P` and `AffineTargetMorphismProperty` in `StableUnderAffineBaseChange.stableUnderBaseChange`, but I am happy to just assume `(of P).StableUnderBaseChange` in this lemma instead. What do you think?

Reply:
I think it makes sense to use `(of P).StableUnderBaseChange` directly.

Reply:
Okay, I removed that part.